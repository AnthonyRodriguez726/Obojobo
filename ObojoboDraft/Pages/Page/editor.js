import React from 'react'

import ActionButton from '../../Chunks/ActionButton/editor'
import Break from '../../Chunks/Break/editor'
import Code from '../../Chunks/Code/editor'
import Figure from '../../Chunks/Figure/editor'
import Heading from '../../Chunks/Heading/editor'
import IFrame from '../../Chunks/IFrame/editor'
import List from '../../Chunks/List/editor'
import MathEquation from '../../Chunks/MathEquation/editor'
import Table from '../../Chunks/Table/editor'
import Text from '../../Chunks/Text/editor'
import YouTube from '../../Chunks/YouTube/editor'
import QuestionBank from '../../Chunks/QuestionBank/editor'
import Question from '../../Chunks/Question/editor'
import MCAssessment from '../../Chunks/MCAssessment/editor'
import MCChoice from '../../Chunks/MCAssessment/MCChoice/editor'
import MCAnswer from '../../Chunks/MCAssessment/MCAnswer/editor'
import MCFeedback from '../../Chunks/MCAssessment/MCFeedback/editor'
import DefaultNode from '../../../src/scripts/oboeditor/components/default-node'

const PAGE_NODE = 'ObojoboDraft.Pages.Page'

const nodes = {
	'ObojoboDraft.Chunks.ActionButton': ActionButton,
	'ObojoboDraft.Chunks.Break': Break,
	'ObojoboDraft.Chunks.Code': Code,
	'ObojoboDraft.Chunks.Figure': Figure,
	'ObojoboDraft.Chunks.Heading': Heading,
	'ObojoboDraft.Chunks.IFrame': IFrame,
	'ObojoboDraft.Chunks.List': List,
	'ObojoboDraft.Chunks.MathEquation': MathEquation,
	'ObojoboDraft.Chunks.Table': Table,
	'ObojoboDraft.Chunks.Text': Text,
	'ObojoboDraft.Chunks.YouTube': YouTube,
	'ObojoboDraft.Chunks.QuestionBank': QuestionBank,
	'ObojoboDraft.Chunks.Question': Question,
	'ObojoboDraft.Chunks.MCAssessment': MCAssessment,
	'ObojoboDraft.Chunks.MCAssessment.MCChoice': MCChoice,
	'ObojoboDraft.Chunks.MCAssessment.MCAnswer': MCAnswer,
	'ObojoboDraft.Chunks.MCAssessment.MCFeedback': MCFeedback,
}

const Node = props => {
	return (
		<div {...props.attributes}>{props.childen}</div>
	)
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	if (node.data) json.content = node.data.get('content') || {}
	json.children = []

	node.nodes.forEach(child => {
		if(nodes.hasOwnProperty(child.type)){
			json.children.push(nodes[child.type].helpers.slateToObo(child))
		} else {
			json.children.push(DefaultNode.helpers.slateToObo(child))
		}
	})

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type
	json.data = { content: node.content }
	json.nodes = []

	node.children.forEach(child => {
		// If the current Node is a registered OboNode, use its custom converter
		if(nodes.hasOwnProperty(child.type)){
			json.nodes.push(nodes[child.type].helpers.oboToSlate(child))
		} else {
			json.nodes.push(DefaultNode.helpers.oboToSlate(child))
		}
	})

	return json
}

const plugins = {
	renderNode(props) {
		switch (props.node.type) {
			case PAGE_NODE:
				return <Node {...props} />
		}
	}
}

const PageNode = {
	components: {
		Node
	},
	helpers: {
		slateToObo,
		oboToSlate,
	}
}

export default PageNode
