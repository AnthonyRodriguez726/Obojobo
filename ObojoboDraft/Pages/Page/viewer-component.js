import './viewer-component.scss';

import Common from 'Common';
import Viewer from 'Viewer';

let { OboComponent } = Common.components;
let { NavUtil } = Viewer.util;

export default class Page {
	componentWillReceiveProps(nextProps) {
		if (nextProps.moduleData.navState.navTargetId !== this.props.moduleData.navState.navTargetId) {
			return NavUtil.setFlag(this.props.moduleData.navState.navTargetId, 'visited', true);
		}
	}

	render() {
		return <OboComponent
			model={this.props.model}
			moduleData={this.props.moduleData}
			className="obojobo-draft--pages--page"
		>
			{
				this.props.model.children.models.map(((child, index) => {
					let Component = child.getComponentClass();

					return <Component key={index} model={child} moduleData={this.props.moduleData} />;
				}))
			}
		</OboComponent>;
	}
}