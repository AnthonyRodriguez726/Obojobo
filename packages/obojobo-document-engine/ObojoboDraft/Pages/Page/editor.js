import React from 'react'
import { Block } from 'slate'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

import ActionButton from '../../Chunks/ActionButton/editor'
import Break from '../../Chunks/Break/editor'
import Code from '../../Chunks/Code/editor'
import Figure from '../../Chunks/Figure/editor'
import Heading from '../../Chunks/Heading/editor'
import IFrame from '../../Chunks/IFrame/editor'
import List from '../../Chunks/List/editor'
import MathEquation from '../../Chunks/MathEquation/editor'
import Table from '../../Chunks/Table/editor'
import Text from '../../Chunks/Text/editor'
import YouTube from '../../Chunks/YouTube/editor'
import QuestionBank from '../../Chunks/QuestionBank/editor'
import Question from '../../Chunks/Question/editor'
import MCAssessment from '../../Chunks/MCAssessment/editor'
import MCChoice from '../../Chunks/MCAssessment/MCChoice/editor'
import MCAnswer from '../../Chunks/MCAssessment/MCAnswer/editor'
import MCFeedback from '../../Chunks/MCAssessment/MCFeedback/editor'
import DefaultNode from '../../../src/scripts/oboeditor/components/default-node'
import Component from '../../../src/scripts/oboeditor/components/editor-component'

const TEXT_NODE = 'ObojoboDraft.Chunks.Text'
const PAGE_NODE = 'ObojoboDraft.Pages.Page'

const nodes = {
	'ObojoboDraft.Chunks.ActionButton': ActionButton,
	'ObojoboDraft.Chunks.Break': Break,
	'ObojoboDraft.Chunks.Code': Code,
	'ObojoboDraft.Chunks.Figure': Figure,
	'ObojoboDraft.Chunks.Heading': Heading,
	'ObojoboDraft.Chunks.IFrame': IFrame,
	'ObojoboDraft.Chunks.List': List,
	'ObojoboDraft.Chunks.MathEquation': MathEquation,
	'ObojoboDraft.Chunks.Table': Table,
	'ObojoboDraft.Chunks.Text': Text,
	'ObojoboDraft.Chunks.YouTube': YouTube,
	'ObojoboDraft.Chunks.QuestionBank': QuestionBank,
	'ObojoboDraft.Chunks.Question': Question,
	'ObojoboDraft.Chunks.MCAssessment': MCAssessment,
	'ObojoboDraft.Chunks.MCAssessment.MCChoice': MCChoice,
	'ObojoboDraft.Chunks.MCAssessment.MCAnswer': MCAnswer,
	'ObojoboDraft.Chunks.MCAssessment.MCFeedback': MCFeedback
}

const Node = props => {
	return <div className={'page-editor'}>{props.children}</div>
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	if (node.data) json.content = node.data.get('content') || {}
	json.children = []

	node.nodes.forEach(child => {
		if (nodes.hasOwnProperty(child.type)) {
			json.children.push(nodes[child.type].helpers.slateToObo(child))
		} else {
			json.children.push(DefaultNode.helpers.slateToObo(child))
		}
	})

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type
	json.data = { content: node.content }
	json.nodes = []

	node.children.forEach(child => {
		json.nodes.push(Component.helpers.oboToSlate(child))
	})

	return json
}

const plugins = {
	renderNode(props) {
		switch (props.node.type) {
			case PAGE_NODE:
				return <Node {...props} {...props.attributes} />
		}
	},
	schema: {
		blocks: {
			'ObojoboDraft.Pages.Page': {
				nodes: [{ match: [ { type: 'oboeditor.component' } ], min: 1 }],
				normalize: (change, error) => {
					const { node, child, index } = error
					switch (error.code) {
						case CHILD_REQUIRED: {
							const block = Block.create({
								object: 'block',
								type: 'oboeditor.component'
							})
							return change.insertNodeByKey(node.key, index, block)
						}
						case CHILD_TYPE_INVALID: {
							const block = Block.create({
								object: 'block',
								type: 'oboeditor.component'
							})
							return change.wrapBlockByKey(child.key, block)
						}
					}
				}
			}
		}
	}
}

const PageNode = {
	components: {
		Node
	},
	helpers: {
		slateToObo,
		oboToSlate
	},
	plugins
}

export default PageNode
