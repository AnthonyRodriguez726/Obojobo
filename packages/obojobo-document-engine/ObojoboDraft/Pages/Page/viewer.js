import Common from 'Common'
import ViewerComponent from './viewer-component'

Common.Registry.registerModel('ObojoboDraft.Pages.Page', {
	componentClass: ViewerComponent,
	default: true,
	type: 'page',
	getNavItem(model) {
		let label

		if (model.title) {
			label = model.title
		} else {
			const pages = model.parent.children.models.filter(
				child => child.get('type') === 'ObojoboDraft.Pages.Page'
			)
			label = `Page ${pages.indexOf(model) + 1}`
		}

		return {
			type: 'link',
			label,
			path: [label.toLowerCase().replace(/ /g, '-')],
			showChildren: false
		}
	}
})
