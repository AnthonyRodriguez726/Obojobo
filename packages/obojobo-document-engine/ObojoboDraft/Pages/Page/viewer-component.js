import './viewer-component.scss'

import Common from 'Common'
import Viewer from 'Viewer'

let { OboComponent } = Common.components

class PageRaw extends React.Component {
	componentWillReceiveProps(nextProps) {
		// @TODO: redux - move this to where the page is actually changed?
		if(nextProps.navTargetId !== this.props.navTargetId){
			this.props.markPageVisited(this.props.navTargetId)
		}
	}

	render() {
		return (
			<OboComponent
				model={this.props.model}
				moduleData={this.props.moduleData}
				className="obojobo-draft--pages--page"
			>
				{this.props.model.children.models.map((child, index) => {
					let Component = child.getComponentClass()

					return <Component key={index} model={child} moduleData={this.props.moduleData} />
				})}
			</OboComponent>
		)
	}
}

let { connect } = Viewer.redux
let { setFlag } = Viewer.redux.NavActions

// Connect to the redux store
const mapStateToProps = (state, ownProps) => ({
	navTargetId: state.nav.navTargetId
})

const mapDispatchToProps = (dispatch, ownProps) => ({
	markPageVisited: (itemId) => { dispatch(setFlag(itemId, 'visited', true))}
})

const Page = connect(mapStateToProps, mapDispatchToProps)(PageRaw)

export default Page
