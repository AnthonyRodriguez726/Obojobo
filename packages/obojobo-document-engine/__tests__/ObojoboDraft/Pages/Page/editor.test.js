import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

import Common from 'Common'
import Page from '../../../../ObojoboDraft/Pages/Page/editor'

const PAGE_NODE = 'ObojoboDraft.Pages.Page'

jest.mock('Common', () => ({
	Registry: {
		registerModel: jest.fn()
	}
}))

describe('Page editor', () => {
	test('plugins.renderNode renders a solution when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: PAGE_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Page.plugins.renderNode(props)).toMatchSnapshot()
	})

	test('plugins.schema.normalize fixes invalid children', () => {
		const change = {
			removeNodeByKey: jest.fn(),
			insertNodeByKey: jest.fn()
		}

		change.withoutNormalization = jest.fn().mockImplementationOnce(funct => {
			funct(change)
		})

		Page.plugins.schema.blocks[PAGE_NODE].normalize(change, {
			code: CHILD_TYPE_INVALID,
			node: { key: 'mockNode' },
			child: { key: 'mockKey' },
			index: null
		})

		expect(change.insertNodeByKey).toHaveBeenCalled()
	})

	test('plugins.schema.normalize adds required children', () => {
		const change = {
			insertNodeByKey: jest.fn()
		}

		Page.plugins.schema.blocks[PAGE_NODE].normalize(change, {
			code: CHILD_REQUIRED,
			node: { key: 'mockKey' },
			child: null,
			index: 0
		})

		expect(change.insertNodeByKey).toHaveBeenCalled()
	})

	test('getNavItem returns expected object', () => {
		const pageMock = Common.Registry.registerModel.mock.calls[0][1]

		const model = {
			parent: {
				children: {
					models: [{ get: () => true }]
				}
			},
			title: 'Test Title'
		}

		expect(pageMock.getNavItem(model)).toEqual({
			type: 'link',
			label: 'Test Title',
			path: ['test-title'],
			showChildren: false
		})

		model.title = null
		expect(pageMock.getNavItem(model)).toEqual({
			type: 'link',
			label: 'Page 0',
			path: ['page-0'],
			showChildren: false
		})
	})
})
