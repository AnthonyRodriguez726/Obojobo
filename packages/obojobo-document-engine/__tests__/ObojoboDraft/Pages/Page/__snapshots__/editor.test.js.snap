// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Node builds the expected component 1`] = `
<div
  className="page-editor"
/>
`;

exports[`Page oboToSlate converts an OboNode to a Slate node 1`] = `
Object {
  "data": Object {
    "content": undefined,
  },
  "key": "mockKey",
  "nodes": Array [
    undefined,
    Object {
      "data": Object {
        "content": undefined,
      },
      "key": undefined,
      "object": "block",
      "type": "NotADefinedNode",
    },
  ],
  "object": "block",
  "type": "mockType",
}
`;

exports[`Page plugins.renderNode renders a solution when passed 1`] = `
<Node
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Pages.Page",
    }
  }
/>
`;

exports[`Page slateToObo converts a Slate node to an OboNode 1`] = `
Object {
  "children": Array [
    undefined,
    Object {
      "children": Array [],
      "id": undefined,
      "type": "NotADefinedNode",
    },
  ],
  "content": Object {},
  "id": "mockKey",
  "type": "mockType",
}
`;

exports[`Page slateToObo converts a Slate node to an OboNode with no content 1`] = `
Object {
  "children": Array [
    undefined,
    Object {
      "children": Array [],
      "id": undefined,
      "type": "NotADefinedNode",
    },
  ],
  "id": "mockKey",
  "type": "mockType",
}
`;
